packages/puppeteer-core/src/common/PSelectorParser.ts

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright 2023 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+type CSSSelector = string;
+
+export type PSelector =
+  | {
+      name: string;
+      value: string;
+    }
+  | CSSSelector;
+
+const PUPPETEER_FUNCTION_TOKEN = /^:-p-([-a-zA-Z_]+)\(/;
+
+class PSelectorParser {
```

Comment:
let's cover this with a unit-level tests? I think there a few edge cases that might be interesting and that are not covered by e2e.

Reply:
Could you provide these cases you are curious about? E2E should cover every case.